### PR TITLE
rwkv6: add wkv6 support for Vulkan backend

### PR DESCRIPTION
**Pull request by zhiyuan1i**
*Make sure to read the [contributing guidelines](https://github.com/ggerganov/llama.cpp/blob/master/CONTRIBUTING.md) before submitting a PR*

This PR aims to enhance support for RWKV6 models, it enables wkv6 at Vulkan backend.

### CODE REVIEW

File: ggml/src/ggml-vulkan/ggml-vulkan.cpp [hunk 1]
```diff
@@ -7724,6 +7838,7 @@ static bool ggml_backend_vk_device_supports_op(ggml_backend_dev_t dev, const ggm
         case GGML_OP_IM2COL:
         case GGML_OP_TIMESTEP_EMBEDDING:
         case GGML_OP_POOL_2D:
+        case GGML_OP_RWKV_WKV6:
```
Comment (jeffbolznv) [inline ggml/src/ggml-vulkan/ggml-vulkan.cpp:7841]: Does this need to check `src0->type == GGML_TYPE_F32 && dst->type == GGML_TYPE_F32` to match get_pipeline?

File: ggml/src/ggml-vulkan/ggml-vulkan.cpp [hunk 2]
```diff
@@ -5424,6 +5439,98 @@ static void ggml_vk_div(ggml_backend_vk_context * ctx, vk_context& subctx, const
...
+
+    ggml_backend_vk_buffer_context * dst_buf_ctx = (ggml_backend_vk_buffer_context *)dst->buffer->context;
+    ggml_backend_vk_buffer_context * k_buf_ctx = (ggml_backend_vk_buffer_context *)k->buffer->context;
+    ggml_backend_vk_buffer_context * v_buf_ctx = (ggml_backend_vk_buffer_context *)v->buffer->context;
+    ggml_backend_vk_buffer_context * r_buf_ctx = (ggml_backend_vk_buffer_context *)r->buffer->context;
+    ggml_backend_vk_buffer_context * tf_buf_ctx = (ggml_backend_vk_buffer_context *)tf->buffer->context;
+    ggml_backend_vk_buffer_context * td_buf_ctx = (ggml_backend_vk_buffer_context *)td->buffer->context;
+    ggml_backend_vk_buffer_context * state_buf_ctx = (ggml_backend_vk_buffer_context *)state->buffer->context;
+
+    vk_buffer d_D = dst_buf_ctx->dev_buffer;
```
Comment (jeffbolznv) [inline ggml/src/ggml-vulkan/ggml-vulkan.cpp:5474]: Does this need to handle uma?

File: ggml/src/ggml-vulkan/ggml-vulkan.cpp [hunk 3]
```diff
@@ -5424,6 +5439,98 @@ static void ggml_vk_div(ggml_backend_vk_context * ctx, vk_context& subctx, const
...
+    vk_buffer d_K = k_buf_ctx->dev_buffer;
+    vk_buffer d_V = v_buf_ctx->dev_buffer;
+    vk_buffer d_R = r_buf_ctx->dev_buffer;
+    vk_buffer d_TF = tf_buf_ctx->dev_buffer;
+    vk_buffer d_TD = td_buf_ctx->dev_buffer;
+    vk_buffer d_State = state_buf_ctx->dev_buffer;
+
+    const uint64_t k_offset = vk_tensor_offset(k);
```
Comment (jeffbolznv) [inline ggml/src/ggml-vulkan/ggml-vulkan.cpp:5482]: add `->view_offset` to these?

File: ggml/src/ggml-vulkan/vulkan-shaders/wkv6.comp [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+void main() {
+    const uint head_size = BLOCK_SIZE;
+    const uint batch_id = gl_WorkGroupID.x / H;
+    const uint head_id = gl_WorkGroupID.x % H;
+    const uint tid = gl_LocalInvocationID.x;
+
+    const uint state_size = C * head_size;
+    const uint n_seq_tokens = T / B;
+
+    if (tid >= head_size || batch_id >= B || head_id >= H) {
```
Comment (jeffbolznv) [inline ggml/src/ggml-vulkan/vulkan-shaders/wkv6.comp:32]: You can remove `tid >= head_size || `, it should never be true.

File: ggml/src/ggml-vulkan/vulkan-shaders/wkv6.comp [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+        return;
+    }
+
+    A_TYPE state[BLOCK_SIZE];
+    for (uint i = 0; i < head_size; i++) {
```
Comment (jeffbolznv) [inline ggml/src/ggml-vulkan/vulkan-shaders/wkv6.comp:37]: Please add `[[unroll]]` for the loops with constant iteration count.

**Review by jeffbolznv — COMMENTED** [review]
I'm not familiar with rwkv, but the shader code appears to correctly match CUDA.

**Comment by jeffbolznv** [comment]
Thanks for the contribution. I'll try to test this locally tomorrow.

File: ggml/src/ggml-vulkan/ggml-vulkan.cpp (hunk 3 above)
Comment (MollySophia) [reply ggml/src/ggml-vulkan/ggml-vulkan.cpp:5482]: That's true. I had these ->view_offset missings in my head. But somehow I forgot to add them when commiting the code. Please help add them @uniartisan 
Thanks!

File: ggml/src/ggml-vulkan/ggml-vulkan.cpp (hunk 1 above)
Comment (uniartisan) [reply ggml/src/ggml-vulkan/ggml-vulkan.cpp:7841]: I'm learning from GGML_OP_POOL_2D to write it, like GGML_OP_POOL_2D, and here is the code on 5,000 lines:
```
    case GGML_OP_RWKV_WKV6:
        if (src0->type == GGML_TYPE_F32 && dst->type == GGML_TYPE_F32) {
            return ctx->device->pipeline_rwkv_wkv6_f32;
        }
        return nullptr;
    case GGML_OP_LEAKY_RELU:
        if (src0->type == GGML_TYPE_F32 && dst->type == GGML_TYPE_F32) {
            return ctx->device->pipeline_leaky_relu_f32;
        }
        return nullptr;
```
Therefore I think there is no need in line 7841.

File: ggml/src/ggml-vulkan/ggml-vulkan.cpp (hunk 3 above)
Comment (uniartisan) [reply ggml/src/ggml-vulkan/ggml-vulkan.cpp]: https://github.com/ggerganov/llama.cpp/pull/10829/files#diff-35a5049d5eebe22eda1e0d661bd87639b31aafeba62deeaaaca9c13ec3e71d11R5516

File: ggml/src/ggml-vulkan/ggml-vulkan.cpp (hunk 1 above)
Comment (jeffbolznv) [reply ggml/src/ggml-vulkan/ggml-vulkan.cpp:7877]: Looks like only F32 is supported in ggml, so I think it's OK to not add the condition in supports_op.

**Review by jeffbolznv — APPROVED** [review]
Thanks for making the changes. LGTM. I ran the backend tests on RTX 4070 and they passed.

**Review by 0cc4m — APPROVED** [review]
Thank you, great work!

**Pull request merged**